CIP-????/README.md

```diff
@@ -0,0 +1,471 @@
+
+As discussed above, both lifting and unlifting are currently very tricky or impossible.
+
+## Specification
+
+This specification has two parts: firstly a change to Plutus Core, and secondly a change to the ledger interface to scripts for a new Plutus language version.
+
+### 1. Plutus Core Changes
+
+The following new term constructors are added to the Plutus Core language:[^typed-plutus-core]
+```
+-- Packs the fields (ts) into a construtor value tagged with i
+(constr i ts)
+
+-- Inspects the tag on t and passes its fields to the corresponding case branch
+(case t cs)
+```
+
+[^typed-plutus-core]: See Appendix 1 for changes to Typed Plutus Core.
+
+The small-step reduction rules for these terms are:
+
+$$
+\mathrm{CONSTR\_SUB}\frac{t_i \rightarrow t'_i}{(\mathrm{constr}\ e\ t_1 \ldots t_n) \rightarrow (\mathrm{constr}\ e\ t'_1 \ldots t'_n)}
+$$
+
+$$
+\mathrm{CASE\_SUB}\frac{t \rightarrow t'}{(\mathrm{case}\ t\ c_1 \ldots c_m) \rightarrow (\mathrm{case}\ t'\ c_1 \ldots c_m)}
+$$
+
+$$
+\mathrm{CASE\_EVAL}\frac{e \leq m}{(\mathrm{case}\ (\mathrm{constr}\ e\ t_1 \ldots t_n)\ c_1 \ldots c_m) \rightarrow [c_e\ t_1 \ldots t_n]}
+$$
+
+$$
+\mathrm{CASE\_ERR}\frac{e \gt m}{(\mathrm{case}\ (\mathrm{constr}\ e\ t_1 \ldots t_n)\ c_1 \ldots c_m) \rightarrow (\mathrm{error})}
+$$
+
+Note that `CASE_SUB` does not reduce branches and `CASE_EVAL` preserves only the selected case branch.
+This means that only the chosen case branch is evaluated, and so `case` is _not_ strict in the other branches.[^strict]
+
+[^strict]: See 'Why is case not stict?' for more discussion.
+
+A new Plutus Core minor language version V is added to indicate support for these terms.
+They are illegal before that version.
+
+#### Costing
+
+All steps in the CEK machine have costs, all of which are constant.
+This will be new costs for the steps for evaluating `constr` and `case`.
+
+There is a potential problem because `constr` and `case` have a variable number of children, unlike all the existing constructs. The risk is that we could end up doing a linear amount of work but only paying a constant cost.
+
+However:
+- `constr`s arguments are all evaluated in turn, so we are sure to pay a linear price.
+- `case`'s branches are _not_ all evaluated, but the only place we could do a linear amount work is in selecting the chosen branch. But this can be avoided in the implementation by storing the cases in a datastructure with constant-time indexing.
+
+### 2. Ledger Changes
+
+A new Plutus ledger language L is added to the ledger.
+The Plutus Core language version V is both allowed and required in L (since the changes to the ledger-script interface will require sums-of-products). 
+Optionally, we may choose to backport support for the Plutus Core language version V to earlier Plutus ledger languages, since it only adds more features.
+
+The ledger-script interface changes in L as follows:
+- Instead of the script context argument being a Data object, it is now a Plutus Core term.
```

Comment:
What will happen to builtins which are taking a `Data` argument right now? For example `serialiseData`

Will users be able to convert these terms into `Data` themselves?

Reply:
Nothing happens to `serialiseData`. 

Let me guess, your real question is "how will this affect the trick Hydra is using to get fast hashing of parts of the script context by using `serialiseData`?". The answer to that is that it won't work because the script context won't be made up of `Data` any more. You could convert it _back_ to `Data`, but that might just eat all your performance improvements.

I'm not sure what to do about that, tbh.

Reply:
> I'm not sure what to do about that, tbh.

A builtin is able to call `asConstant` on the given value, we could add `asConstr` alongside and recurse within the freshly added `serializeSumOfProd` builtin.

Reply:
Is there any reason to keep `Data` in UPLC at all? It's not apparent to me from the CIP, but datums and redeemers could also be made to use this, no?

Reply:
Ah yes, I should have discussed this in the proposal body, I'll add something. Basically, my argument is that datums and redeemers are _different_ to the script context, because they're actually user-facing. They appear in transactions themselves, users want to be able to construct them even on the command-line, etc. So there's a real benefit there to having them be a separate, restricted format, rather than arbitrary Plutus Core terms. While this proposal _does_ make Plutus Core terms less opaque, they're still not guaranteed to be as easily inspectable as `Data`.

Reply:
@michaelpj What I meant was, while datums and redeemers could be kept as "Data" from the perspective of the ledger, I don't see any reason why they can't be represented using UPLC SoPs? `Data` has 5 constructors, each would be encoded using `constr`. `Constr` would rather than including an integer index and a list of data be changed to include a single SoP encoded using `constr`, so you would have e.g. `constr 0 (constr 0 x y z)` for a datum that is a product of three fields `x y z`.

Reply:
It's possible there are more efficient encodings. You could feasible "inline" the index of the wrapped SoP into the index of the Data SoP, such that anything beyond 4 is user-defined.

Reply:
Interesting. I can't decide if I love this idea or hate it :joy: 

If we did this, we'd need a specified conversion `data2Term :: Data -> Term`, but then we could e.g. use that same translation for the ledger. That means we would say something like "the ledger still converts the script context to Data as before, the change is just that we turn all the arguments into Terms with data2Term rather than turning them into builtin constants". So in some ways it gives us a nice general story :thinking:

Reply:
@L-as I added this as an Alternative, since I'm still unsure if we should do it but I wanted to record it.